schema/context.go

```diff
@@ -0,0 +1,37 @@
+package schema
+
+import (
+	"context"
+)
+
+type bodyExtCtxKey struct{}
+
+type bodyActiveCountCtxKey struct{}
+
+func WithActiveCount(ctx context.Context) context.Context {
+	return context.WithValue(ctx, bodyActiveCountCtxKey{}, true)
+}
+
+func ActiveCountFromContext(ctx context.Context) bool {
+	return ctx.Value(bodyActiveCountCtxKey{}) != nil
+}
+
+type bodyActiveForEachCtxKey struct{}
+
+func WithActiveForEach(ctx context.Context) context.Context {
+	return context.WithValue(ctx, bodyActiveForEachCtxKey{}, true)
+}
+
+func ActiveForEachFromContext(ctx context.Context) bool {
```

Comment:
This function is only used within the semantic tokens:
https://github.com/hashicorp/hcl-lang/blob/6bf6bd83e2bc2b61aa8df2aab8ce4e60c0776852/decoder/semantic_tokens.go#L176-L193

But with the matching of references and origins, it might not be required anymore to detect a valid traversal step.